### PR TITLE
VZ-7637 - Enable use of Rancher proxy URL in kubeconfig used by managed clusters to connect to admin cluster.

### PR DESCRIPTION
**Pull request by desagar**
Before DNS updates, copy the Rancher ingress and TLS secret(s) so that managed clusters can use the old DNS name to access the admin cluster until the update is propagated.

### CODE REVIEW

File: platform-operator/controllers/verrazzano/reconcile/rancher.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+			ctx.Log().Debugf("Rancher ingress host %s has DNS suffix %s, skipping copy", ing.Spec.TLS[0].Hosts[0], dnsSuffix)
+			continue
+		}
+
+		newIngressTLSList := r.createIngressCertSecretCopies(ctx, ing)
+		newIngressNSN := types.NamespacedName{Name: "vz-" + ing.Name, Namespace: ingressName.Namespace}
+		err = r.createIngressCopy(newIngressNSN, ing, newIngressTLSList)
+		if err != nil {
+			ctx.Log().Infof("Failed to create a copy of Rancher ingress %s/%s - create ingress failed: %v", ingressName.Namespace, ingressName.Name, err)
+			continue
```
Comment (davisgomes) [inline platform-operator/controllers/verrazzano/reconcile/rancher.go:63]: What is the justification for ignoring this error?

Comment (davisgomes) [reply platform-operator/controllers/verrazzano/reconcile/rancher.go:63]: And the ones prior

File: platform-operator/controllers/verrazzano/reconcile/rancher_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+				asserts.Equal(t, ing.Spec.TLS[i].SecretName, "vz-"+tls.SecretName)
+			}
+		}
+	}
+
+	secretList := corev1.SecretList{}
+	err = c.List(context.TODO(), &secretList)
+	asserts.NoError(t, err)
+	asserts.Equal(t, 4, len(secretList.Items))
+}
```
Comment (davisgomes) [inline platform-operator/controllers/verrazzano/reconcile/rancher_test.go:104]: Could there be a test here that verifies that everything runs smoothly (or does what you expect) if some of the expected resources are missing?

File: platform-operator/controllers/verrazzano/reconcile/rancher.go (hunk 1 above)
Comment (desagar) [reply platform-operator/controllers/verrazzano/reconcile/rancher.go:63]: I think the idea I had was that this was not something that should fail the Rancher install/update and hence the VZ install/update. This copying is needed only in a subset of cases and there is a workaround i.e. manually reapplying the registration.
Of course, the downside is that diagnosability when multicluster does not work is made a bit more difficult. I am open to opinions on whether we should make this a failing change.

Comment (fredtibbitts) [reply platform-operator/controllers/verrazzano/reconcile/rancher.go:63]: I had thought about this too, we'd have to make `beforeInstallComponents` return an error and then requeue and not proceed with installing components until it passes. I've gone back and forth, but I think I'm leaning towards making this a "best effort" and not fail the install. We can always make it more restrictive in the future.

File: tests/e2e/update/certac/certac_test.go [hunk 3]
```diff
@@ -177,7 +217,88 @@ func verifyRegistration() {
...
+
+// reapplyManagedClusterRegManifest reapplies the registration manifest on managed clusters. For
+// self-signed certs, this manual step is expected of users so that the admin kubeconfig used by the
+// managed clusters can be reliably updated to use the new CA cert. Otherwise intermittent timing
+// related failures may occur
+func reapplyManagedClusterRegManifest(newCACert string) {
+	// CAUpdateTime := time.Now()
+	for _, managedCluster := range managedClusters {
+		waitForManifestSecretUpdated(managedCluster.Name, newCACert)
+		// waitForVMCReadyAfterTime(managedCluster.Name, CAUpdateTime)
```
Comment (fredtibbitts) [inline tests/e2e/update/certac/certac_test.go:233]: Nit, remove commented code (and the one a few lines above).

File: tests/e2e/update/certac/certac_test.go [hunk 4]
```diff
@@ -177,7 +217,88 @@ func verifyRegistration() {
...
+		gomega.Eventually(func() error {
+			reg, err := adminCluster.GetManifest(managedCluster.Name)
+			if err != nil {
+				pkg.Log(pkg.Error, fmt.Sprintf("could not get manifest for managed cluster %v error: %v", managedCluster.Name, err))
+				return err
+			}
+			pkg.Log(pkg.Info, fmt.Sprintf("Reapplying registration manifest for managed cluster %s", managedCluster.Name))
+			managedCluster.Apply(reg)
+			return nil
+		}).WithTimeout(waitTimeout).WithPolling(pollingInterval).Should(gomega.BeNil(), fmt.Sprintf("Reapply registration manifest failed for cluster %s", managedCluster.Name))
```
Comment (fredtibbitts) [inline tests/e2e/update/certac/certac_test.go:243]: `ShouldNot(HaveOccurred())` reads a little better.

Comment (desagar) [reply tests/e2e/update/certac/certac_test.go]: thanks, missed this one. I had updated it elsewhere. Fixed.

File: tests/e2e/update/certac/certac_test.go (hunk 3 above)
Comment (desagar) [reply tests/e2e/update/certac/certac_test.go]: done

File: platform-operator/controllers/verrazzano/reconcile/rancher_test.go (hunk 2 above)
Comment (desagar) [reply platform-operator/controllers/verrazzano/reconcile/rancher_test.go:179]: I'm not sure what you mean by missing resources, but I added one for when the copy already exists and made sure it was updated without errors. (e.g. when you do 2 DNS updates in succession).

Comment (davisgomes) [reply platform-operator/controllers/verrazzano/reconcile/rancher_test.go:179]: I wasn't sure what feasible errors you would come across, but that seems reasonable to me.

**Review by davisgomes — APPROVED** [review]

**Review by fredtibbitts — APPROVED** [review]

**Review by davisgomes — APPROVED** [review]

**Pull request merged**